internals/overlord/checkstate/manager.go

```diff
@@ -48,42 +100,99 @@ func (m *CheckManager) NotifyCheckFailed(f FailureFunc) {
 
 // PlanChanged handles updates to the plan (server configuration),
 // stopping the previous checks and starting the new ones as required.
-func (m *CheckManager) PlanChanged(p *plan.Plan) {
-	m.mutex.Lock()
-	defer m.mutex.Unlock()
+func (m *CheckManager) PlanChanged(newPlan *plan.Plan) {
+	m.state.Lock()
+	defer m.state.Unlock()
+
+	shouldEnsure := false
+	newOrModified := make(map[string]bool)
+	existingChecks := make(map[string]bool)
+
+	// Abort all currently-running checks that have been removed or modified.
+	for _, change := range m.state.Changes() {
+		switch change.Kind() {
+		case performCheckKind, recoverCheckKind:
+			if change.IsReady() {
+				// Skip check changes that have finished already.
+				continue
+			}
+			details := mustGetCheckDetails(change)
+			var oldConfig *plan.Check
+			if change.Kind() == performCheckKind {
+				oldConfig = m.state.Cached(performConfigKey{change.ID()}).(*plan.Check)
+			} else {
+				oldConfig = m.state.Cached(recoverConfigKey{change.ID()}).(*plan.Check)
+			}
+			existingChecks[oldConfig.Name] = true
+
+			newConfig, inNew := newPlan.Checks[details.Name]
+			if inNew {
+				// TODO: add test of service context changing
+				merged := mergeServiceContext(newPlan, newConfig)
+				if reflect.DeepEqual(oldConfig, merged) {
+					// Don't restart check if its configuration hasn't changed.
+					continue
+				}
+				// Check is in old and new plans and has been modified.
+				newOrModified[details.Name] = true
+			}
+			change.Abort()
+			m.deleteCheckInfo(details.Name)
+			shouldEnsure = true
+		}
+	}
 
-	logger.Debugf("Configuring check manager (stopping %d, starting %d)",
-		len(m.checks), len(p.Checks))
+	// Also find checks that are new (in new plan but not in old one).
+	for _, config := range newPlan.Checks {
+		if !existingChecks[config.Name] {
+			newOrModified[config.Name] = true
+		}
+	}
 
-	// First stop existing checks.
-	for _, check := range m.checks {
-		check.cancel()
+	// Start new or modified checks.
+	for _, config := range newPlan.Checks {
+		if newOrModified[config.Name] {
+			merged := mergeServiceContext(newPlan, config)
+			changeID := performCheckChange(m.state, merged)
+			m.updateCheckInfo(config, changeID, 0)
+			shouldEnsure = true
+		}
+	}
+	if !m.ensureDone.Load() {
+		// Can't call EnsureBefore before Overlord.Loop is running (which will
+		// call m.Ensure for the first time).
+		return
```

Comment:
So when is the EnsureBefore called if this is false? does it need to be called?

Reply:
In the overlord startup when it loads the plan the first time (I've traced it through). That's what the comment is referring to -- not actually EnsureBefore which is a helper function, but Ensure itself.